File: 18.9/sell_vegetables_ex.py

# score: 50

length = int(input())
second = [int(x) for x in input().split()]

def process(result, idx):
    if idx == 0:
        result.append(1)
    elif idx == 1:
        num = 2*second[idx-1] - result[idx-1]
        while num < 1:
            num += 1
            if num >= 2*(second[idx-1]+1) - result[idx-1]:
                return False
        result.append(num)
    elif idx == length:
        if (result[idx-2]+result[idx-1])//2 != second[idx-1]:
            return False
        else:
            print(' '.join([str(x) for x in result]))
            return True
    else:
        num = 3*second[idx-1] - result[idx-1] - result[idx-2]
        while num < 1:
            num += 1
            if num >= 3*(second[idx-1]+1) - result[idx-1] - result[idx-2]:
                return False
        result.append(num)


    while True:
        if (process(result.copy(), idx+1) == False):
            result[idx] = result[idx]+1
            if idx==1 and (result[idx]>= 2*(second[idx-1]+1) - result[idx-1]):
                return False
            elif idx>1 and (result[idx] >= 3*(second[idx-1]+1) - result[idx-1] - result[idx-2]):
                return False
        else:
            return True


process([], 0)

File: 18.9/sell_vegetables_ex.cpp
#include<vector>
#include<iostream>

using namespace std;

int n;
vector<int> v, result;

bool process(int k) {
    if (k > 1 && k < n) {
        result[k] = v[k-1]*3 - result[k-1] - result[k-2];
        if (result[k] < 1) result[k] = 1;
        while (true) {
            if ((result[k] + result[k-1] + result[k-2])/3 > v[k-1]) return false;

            if (process(k+1)) return true;
            else ++result[k];
        }
        return true;
    }
    else if (k == 0) {
        result[k] = 1;
        while (!process(k+1)) {
            ++result[k];
        }
        return true;
    }
    else if (k == 1) {
        result[k] = v[k-1]*2 - result[k-1];
        if (result[k] < 1) result[k] = 1;
        while (true) {
            if ((result[k] + result[k-1])/2 > v[k-1]) return false;

            if (process(k+1)) return true;
            else ++result[k];
        }
        return true;
    }
    else {
        if ((result[k-2] + result[k-1])/2 == v[k-1]) return true;
        else return false;
    }
}

int main() {
    cin >> n;
    int tmp;
    for (int i = 0; i < n; ++i) {
        cin >> tmp;
        v.push_back(tmp);
        result.push_back(0);
    }
    if (n == 1) {
        cout << v[0] << endl;
        return 0;
    }
    process(0);
    for (int i = 0; i < n; ++i) {
        cout << result[i];
        if (i < n-1) cout << " ";
    }
    cout << endl;
    return 0;
}